Add JSON UO reader for interaction with json data files from input

GiudGiud opened the pull request:
@simopier for the tritium diffusivities I was thinking that

This mirrors the CSV reading capability (though not in those particular functions, that predate the PropertyFileReader format)

@lynnmunday this could be the template for adding the reporter vectors as X Y data like you had for optimization. It would be pretty easy that way

Refs #25058

GiudGiud committed:
Add a JSON reader from the input file
only a scalar getter for now

GiudGiud committed:
Include a more developed getter with a vector of keys

GiudGiud committed:
Reform json contrib for readability

GiudGiud committed:
Add getters for vectors

GiudGiud committed:
Rework and clean parameters 1D tabular functions

GiudGiud committed:
Actual parameter checking in 1D tabular function base class

GiudGiud committed:
Add support for JSON input in 1D tabular function
refs #25058

GiudGiud committed:
Accomodate for late loading of JSON data from UO
refs #25058

GiudGiud committed:
Add useful errors for mis-formatted or mis-read JSON files
refs #25058

GiudGiud committed:
Add test for function loading data from JSON
refs #25058

GiudGiud committed:
Add exception checking for odd parameters for the piecewise tabular 1D functions
Adapt existing exception checking

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1786733/) on bd2cd2a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25568/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1786928/) on 05c4d13 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>81a54c</th>
      <th colspan=4>#25568 05c4d1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.68%</td>
      <td>85.66%</td>
      <td>-0.02%</td>
      <td>71.90%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91590</td>
      <td>91659</td>
      <td>+69</td>
      <td>87</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15308</td>
      <td>15346</td>
      <td>+38</td>
      <td>34</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25568/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25568/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25568/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 71.90% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Add doco for JSONFileReader

GiudGiud commented:
Patches are up for 3 apps.

lindsayad reviewed (changes requested):
Can we get more coverage of the `JSONFileReader`?

GiudGiud committed:
Add a unit test for JSON file reader UO

GiudGiud commented:
Good idea. Added a unit test

lindsayad commented on test/tests/functions/piecewise_constant/tests:
```diff
@@ -1,16 +1,48 @@
+      cli_args = 'Functions/left/x_keys=hello'
+      expect_err = "A parameter was passed for a JSON input option but the \'json_uo\' parameter has "
+                   "not been set"
+      detail = 'parameters meant for loading data from JSON are passed for a function loading data differently,'
```
```suggestion
      detail = 'or if parameters meant for loading data from JSON are passed for a function loading data differently.'
```

lindsayad commented on test/tests/misc/check_error/tests:
```diff
@@ -456,7 +473,8 @@
 
-    requirement = "The system shall report an error when the coupled function does not exist or can not be parsedfor the Piecewise utility."
+    requirement = "The system shall report an error when the coupled function does not exist or can "
+                  "not be parsedfor the Piecewise utility."
```
```suggestion
                  "not be parsed by a piecewise utility."
```

GiudGiud commented on framework/contrib/json/include/nlohmann/json.h:
no I m fixing the docstring because it s not readable

GiudGiud commented on framework/include/functions/PiecewiseTabularBase.h:
```diff
@@ -33,10 +36,16 @@ class PiecewiseTabularBase : public PiecewiseBase
   const bool _has_axis;
   ///@}
 
+  /// Boolean to keep track of whether the data has been loaded
+  bool _raw_data_loaded;
+
 private:
   /// Reads data from supplied CSV file.
   void buildFromFile();
 
+  /// Reads data from supplied JSON reader.
+  void buildFromJSON();
```
it works? we dont want derived classes using these, we're happy with the derived classes using the attributes?

lindsayad replied:
but that's odd right? Methods help encapsulate data. A user of this class could just toggle `_raw_data_loaded` to false willy/nilly irrespective of what "internal" operations have happened in this class

GiudGiud replied:
but not the same data here.
all the data members here are from the input params, while the build holds the raw arrays

we have protected/private wrong all over MOOSE. We use protected way too much when it s commonly adivsed against
https://softwareengineering.stackexchange.com/questions/162643/why-is-clean-code-suggesting-avoiding-protected-variables

GiudGiud replied:
but you have a point, that data should just be private. Hopefully changing the old ones wont force me to go fix too much downstream code

GiudGiud replied:
ok all the others members are used. I moved mine to private as it should

I ll let someone else enforce removing protected data members and add proper getters/setters

lindsayad replied:
Did you move it?

GiudGiud replied:
I think I did?

GiudGiud replied:
it didnt move!

GiudGiud replied:
ok should be good now

GiudGiud commented on framework/include/userobjects/JSONFileReader.h:
```diff
@@ -0,0 +1,80 @@
+
+  /**
+   * Get a scalar in the JSON file/tree
+   * @param scalar_keys the keys in descending order to access the scalar
```
```suggestion
   * @param scalar_keys the keys in descending order to access the scalar in the JSON tree
```

GiudGiud commented on framework/include/userobjects/JSONFileReader.h:
```diff
@@ -0,0 +1,80 @@
+  void getScalar(const std::string & scalar_name, Real & scalar) const;
+
+  /**
+   * Get a scalar in the JSON file/tree
```
```suggestion
   * Get a scalar in the JSON file/tree using the keys in the vector
```

GiudGiud committed:
Address Alex's review

- dont use references for small things
- use a reference for the string
- remove extra code space
- English official language

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1788987/) on bee8ccd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25568/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25568/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 94e06424a4098724b6cb88648d197c3187df9a07`

GiudGiud commented on framework/include/userobjects/JSONFileReader.h:
```diff
@@ -0,0 +1,80 @@
+   * @param scalar_keys the keys in descending order to access the scalar
+   * @param scalar reference to the scalar that will be set with the scalar in the JSON
+   */
+  void getScalar(const std::vector<std::string> & scalar_keys, Real & scalar) const;
```
added more text

GiudGiud commented on framework/include/userobjects/JSONFileReader.h:
```diff
@@ -0,0 +1,80 @@
+  Real getReal(const nlohmann::json & node) const;
+
+  /// Database filename
+  const FileName _filename;
```
I get that
```
/Users/giudgl/projects/moose/framework/src/userobjects/JSONFileReader.C:30:46: error: reference member '_filename' binds to a temporary object whose lifetime would be shorter than the lifetime of the constructed object
  : GeneralUserObject(parameters), _filename(getParam<std::string>("filename"))
```

lindsayad commented on framework/src/userobjects/JSONFileReader.C:
```diff
@@ -0,0 +1,125 @@
+{
+  InputParameters params = GeneralUserObject::validParams();
+  // Add parameters
+  params.addRequiredParam<std::string>("filename", "The path to the file including its name");
```
```suggestion
  params.addRequiredParam<FileName>("filename", "The path to the file including its name");
```

lindsayad commented on framework/src/userobjects/JSONFileReader.C:
```diff
@@ -0,0 +1,125 @@
+}
+
+JSONFileReader::JSONFileReader(const InputParameters & parameters)
+  : GeneralUserObject(parameters), _filename(getParam<std::string>("filename"))
```
```suggestion
  : GeneralUserObject(parameters), _filename(getParam<FileName>("filename"))
```

GiudGiud commented:
ok got it for the conversion creating a temporary

GiudGiud commented on framework/doc/content/source/userobjects/JSONFileReader.md:
```diff
@@ -0,0 +1,18 @@
+# JSONFileReader
+
```
last push was fixing this up a little

GiudGiud commented:
```
/data/civet1/build/moose/framework/src/functions/PiecewiseLinear.C:28:7: error: '_raw_data_loaded' is a private member of 'PiecewiseTabularBase'
  if (_raw_data_loaded)
```
protected is too nice. 

There could be write-only and read-only access with a few keywords in C++. We should not have to triplicate member - getter - setter

GiudGiud committed:
Add support for piecewise coarse linear for JSON

GiudGiud commented:
last commit shows the issue with protected quite well though.
we let people access _raw_x, and now we cant stop people from accessing them before the data is actually loaded.

now if they code well within our standards, they should be asserting the size of raw_x to protect themselves, but we cant enforce that

GiudGiud commented:
patches are up for all apps now
@joshuahansel  for sockeye: https://github.inl.gov/ncrc/sockeye/pull/727

lindsayad commented:
> protected is too nice.
> 
> There could be a read-only 'private' access in C++. It would be nice. We should not have to triplicate member - getter - setter

I totally agree that it can be very convenient to make data protected. It makes writing downstream code much easier, while also making it easier for downstream code to shoot themselves in the foot.

My main point was that having methods private and data protected doesn't make much sense. A downstream class could just reimplement the private method or do whatever they want with the protected data, so declaring the method as `private` in the base class is a farce as far as protecting the base class's state goes.

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/1789816/) on 1914bf4 : invalidated by @GiudGiud

patch propagated

GiudGiud commented:
All patches merged. Thank you for the review!
